### PR TITLE
*: temporarily remove distributed mode

### PR DESCRIPTION
**Pull request by benesch**
Very early on in its life, Materialize learned to support a distributed
mode (#228), powered by Timely's support for scaling computation across
multiple processes. Unfortunately we never found the time to
productionalize this mode, and as a result it is quite fragile. We do
not presently recommend that anyone use distributed mode in production.

The headline problem with distributed mode is that a single transient
network error will cause the entire process to crash. This is true of
the network communications managed by both Timely and Materialize. The
assumption that the network is reliable is baked in pretty deeply and is
therefore not easy to change on short notice.

Even though we discourage use of distributed mode, we carry around a lot
of complexity to support it:

  * The entire `comm` package exists only in service of distributed mode.
    It provides channels that work over TCP, so that the coordinator can
    communicate with workers in different processes.

  * `comm` channels are clunky to use, because you have to initiate a
    TCP connection before you can use them. This is a fallible
    operation, which means code interacting with these channels needs to
    be prepared to handle errors nearly everywhere... of course, none of
    it does, so we just have a lot of `unwraps` lying around.

  * `comm` channels route interthread communication through the kernel,
    which is... certainly not a performance enhancement, though it's
    unclear just how costly it is.

  * `comm` channels use up ports/file descriptors, which are not free
    resources. Users have reported crashes in Materialize due to file
    descriptor exhaustion.

  * Server startup is complicated because it has to handle the case of a
    process that is only hosting Timely workers, and not a coordinator
    or pgwire server or HTTP server.

  * Network configuration is complicated because a Materialize process
    needs to listen for incoming Timely worker traffic from other nodes,
    rather than just incoming user traffic. This increases the surface
    area that needs to be secured when exposing a materialized process
    to a public network.

This commit is a proposal to temporarily remove distributed mode and all
the complexity it entails. The plan is still very much to support a
distributed mode in the future, but to do so when we have the
engineering bandwidth to manage the complexity.

Specifically, this commit:

  * removes the `comm` package,

  * replaces all use of `comm` channels with channels from either
    `tokio` or `crossbeam_channel`, i.e., production-grade channels with
    proper semantics and performance,

  * eliminates the concept of distributed mode from server startup.

The immediate benefit is that we needn't worry about securing Timely's
internal communications in the cloud product. Other short-term benefits
include possible performance speedups (we should measure!) and
simplifications in the coordinator.

Of course, there is a long-term downside here, which is that we will
start to implicitly encode assumptions about the fact that Materialize
is a single-node system into the code. All those assumptions will have
to be rewound when we do pursue distributed mode in the future.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5638)

### CODE REVIEW

**Comment by benesch** [comment]
To attempt to briefly clarify a question that has come up, which is roughly summarized as "why not just fix the issues with `comm`?"

The short answer is that I don't think we have the engineering bandwidth right now. Here are the three big issues.

1. Security. When Materialize is exposed to a public network all network communication needs to be encrypted and authenticated. `comm` does not presently support this. It could be taught to do so without _too_ much trouble, but that is additional code to write, maintain, and audit for security. The latter point—the additional surface area that needs to be audited for security—is the most serious.

2. Reliability. Loopback TCP occasionally fails. Presently that causes Materialize to crash, because nearly every call to `send` or `recv` on a `comm` channel is followed by an unwrap. Sometimes these glitches seem to be weird networking hiccups; other times they are the clear result of a user running out of ports or file descriptors. These problems are fundamentally tractable—we can ask the users to bump their rlimits; multiplex communication onto fewer `comm` channels; sprinkle retry loops into the right places; etc.—but if we switch to shared-memory channels these problems just disappear overnight.

3. Performance. `comm` has a very non optimal implementation for communications between threads hosted by the same process. Those threads share an address space, and ought to be communicating via shared memory (or a channel implementation backed by shared memory). `comm` channels, by contrast, serialize the data, send it through the kernel's loopback TCP interface, and then deserialize it on the other end. That's a lot of unnecessary overhead.

    To fix this issue, we'd need to have swappable implementations of channels for `comm`, and use the fast-local implementation in single node mode. But the current design of `comm` is very tightly tied to the idea of a connection being a byte-oriented network stream where a channel endpoint is an address is some global space (i.e., a hostname and port), and would require a nontrivial amount of reworking to support shared memory channels.

The issue that sparked this PR is (1) which is presently a blocker for the cloud console. I believe that it would only be a few days of work to write the _code_ to fix (1), though there is still the question of whether we're comfortable with the additional threat vector of opening a port that speaks a homegrown protocol. But solving (1), and leaving it at that, is certainly a workable alternative to this PR.

As I looked into this issue (2) was also deeply worrying to me. Certainly in a world where we support distributed mode, communication will need to flow over the network and we will need to solve these port/FD exhaustion problems properly. But investing the time and energy now to shake those bugs out seems like a bit of a misallocation of resources. We're paying the costs of a distributed architecture, but we're not getting much in return because we're not actually recommending that anyone use distributed mode.

To summarize: this PR fixes all three issues in the short term at a long term cost. The alternative take in which we keep `comm` realistically results in only (1) getting fixed, because I don't think we have the bandwidth for addressing (2) or (3) right now.

**Review by frankmcsherry — APPROVED** [review]
This looks good. There are a few things I didn't understand (why we don't need an `Option` in one place, a bool return in another, what `Derivative` means), but none of these were things I would notice reading the code without the diffs.

**Comment by benesch** [comment]
> This looks good. There are a few things I didn't understand (why we don't need an `Option` in one place, a bool return in another, what `Derivative` means), but none of these were things I would notice reading the code without the diffs.

👍🏽 well tysm for taking a look! I can't comment on those `bool`/`Options but `Derivative` is a handy little alternative to `#[derive(Debug)]` that allows you to sprinkle some `#[ignore_this_field]` annotations around. That way you don't have to chuck the baby out with the bathwater if just one field in a struct doesn't impl `Debug`.

**Comment by benesch** [comment]
Ohhkay, into the deep we go!
